kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md

```diff
@@ -221,5 +253,144 @@ module MANAGEDEI
+              ~> #pushBytes(Int2Bytes(NONCE, BE, Unsigned))
+              ~> #getBuffer(TOK_IDX)
+              ~> #appendBytes
+              ~> #getESDTTokenData
+                 ...
+        </instrs>
+        <locals>
+          ...
+          0 |-> <i32> ADDR_IDX
+          1 |-> <i32> TOK_IDX
+          2 |-> <i64> NONCE
+          ...
+        </locals>
+
+    syntax InternalInstr ::= "#getESDTTokenData"
+ // --------------------------------------------
+    rule [getESDTTokenData-nft]:
+        <instrs> #getESDTTokenData
+              => #setBigIntValue(VAL_IDX, VALUE)
+              ~> #setBuffer(PROPS_IDX, PROPS)
+              ~> #setBuffer(HASH_IDX, HASH)
+              ~> #setBuffer(NAME_IDX, NAME)
+              ~> #setBuffer(ATTRS_IDX, ATTRS)
+              ~> #setBuffer(HASH_IDX, HASH)
+              ~> #setBuffer(CREATOR_IDX, CREATOR)
+              ~> #setBigIntValue(ROYL_IDX, ROYL)
+              ~> #writeManagedVecOfManagedBuffers(URIS, URIS_IDX)
+                 ...
+        </instrs>
+        <bytesStack> TOKEN : ADDR : REST => REST </bytesStack>
+        <locals>
+          ...
+          3 |-> <i32> VAL_IDX
+          4 |-> <i32> PROPS_IDX
+          5 |-> <i32> HASH_IDX
+          6 |-> <i32> NAME_IDX
+          7 |-> <i32> ATTRS_IDX
+          8 |-> <i32> CREATOR_IDX
+          9 |-> <i32> ROYL_IDX
+          10 |-> <i32> URIS_IDX
+          ...
+        </locals>
+        <account>
+          <address> ADDR </address>
+          <esdtData>
+            <esdtId> TOKEN </esdtId>
+            <esdtBalance> VALUE </esdtBalance>
+            <esdtMetadata>
+              esdtMetadata( ...
+                name: NAME,
+                creator: CREATOR,
+                royalties: ROYL,
+                hash: HASH,
+                uris: URIS,
+                attributes: ATTRS
+              )
+            </esdtMetadata>
+            <esdtProperties> PROPS </esdtProperties>
+            ...
+          </esdtData>
+          ...
+        </account>
+
+    rule [getESDTTokenData-ft]:
+        <instrs> #getESDTTokenData
+              => #setBigIntValue(VAL_IDX, VALUE)
+              ~> #setBuffer(PROPS_IDX, PROPS)
+                 ...
+        </instrs>
+        <bytesStack> TOKEN : ADDR : REST => REST </bytesStack>
+        <locals>
+          ...
+          3 |-> <i32> VAL_IDX
+          4 |-> <i32> PROPS_IDX
+          ...
+        </locals>
+        <account>
+          <address> ADDR </address>
+          <esdtData>
+            <esdtId> TOKEN </esdtId>
+            <esdtBalance> VALUE </esdtBalance>
+            <esdtMetadata> .esdtMetadata </esdtMetadata>
+            <esdtProperties> PROPS </esdtProperties>
+            ...
+          </esdtData>
+          ...
+        </account>
+
+    // extern long long getESDTLocalRoles(void* context, int32_t tokenIdHandle);
+    rule [getESDTLocalRoles]:
+        <instrs> hostCall("env", "getESDTLocalRoles", [ i32 .ValTypes ] -> [ i64 .ValTypes ])
```

Comment:
How does this rule actually work? And why does it check that there are no esdtTransfers? Should it have a "#getESDTLocalRoles" on the RHS?

Reply:
Sorry, this function is not finished yet. I should have excluded it in the commit.